app/models/package_manager/base/bulk_version_updater.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+module PackageManager
+  class Base
+    # This class is responsible for taking a project and a list of ApiVersions (aka raw versions mapped
+    # to a common set of attributes), and creating Versions for the Project in the database, in a single
+    # query. Because we're using +upsert_all()+, Versions that already exist will just be updated.
+    # This single query avoids all the N+1s involved in a regular batch insert of Versions that we did in the past.
+    class BulkVersionUpdater
+      def initialize(db_project:, api_versions:, repository_source_name:)
+        @db_project = db_project
+        @api_versions = api_versions
+        @repository_source_name = repository_source_name
+      end
+
+      def run!
+        return if @api_versions.empty?
+
+        # create synthetic versions without saving them, so we can get their attributes
+        attrs = @api_versions
+          .map { |api_version| Version.new(api_version.to_version_model_attributes.merge(project: @db_project)) }
+          .each do |v|
+            # these will get merged properly in the upsert_all
+            v.created_at = v.updated_at = Time.current
+            # this value will get merged w/existing in the upsert_all query (see table tests in spec)
+            v.repository_sources = @repository_source_name if @repository_source_name
+            # from Version#before_save
+            v.update_spdx_expression
+            # upsert_all doesn't do validation, so ensure they're valid here.
+            v.validate!
+          end
+          .map { |v| v.attributes.without("id") }
+
+        existing_version_ids = @db_project.versions.all.pluck(:id)
+
+        # TODO: we could do this in batches if performance does not scale well with # of versions.
+        Version.upsert_all(
+          attrs,
+          # handles merging any existing repository_sources with new repository_source (see specs for table tests)
+          # note that timestamps act slightly differently: we'll use the provided updated_at here if it exists, which 
+          # might update updated_at even if no attributes changed. Whereas normally with AR's dirty attributes, timestamps
+          # wouldn't change unless other attributes had changed during the save.
```

Comment:
this is really the only big data change: before, we wouldn't update a Version if its attributes hadn't changed. But now, since we're passing all the Version attributes into one query, we'll take the `updated_at` if it's passed here (which it will be).

Reply:
This wouldn't trigger extra Release changes when we sync, right?

Reply:
nah, for releases that already exist, the `on_duplicate` logic below dictates how we update any info:
* status: pick new over old
* runtime_dependencies_count: pick new over old 
* original_license: pick new over old 
* published_at: pick new over old 
* created_at: use old, and if it isn't set then use new
* updated_at: use new, and if it isn't set then use old
* repository_sources: merge new into old, without duplicates or nulls

Reply:
Updating the updated_at if nothing's changed doesn't seem great, esp since it's used for debugging data issues and is not standard object behavior :/  I would be expecting something like the "last checked at" timestamp to be bumped, but not the updated_at.

Reply:
I'll see what I can do! Might not be too bad to preserve dirty state